File: src/commands/remix.rs
use crate::utils::gemini_imagen_client::GeminiImagenPromptWithImage;
use crate::Data;
use poise::serenity_prelude::{Color, CreateAttachment, CreateEmbed};
use poise::{command, CreateReply};
use tracing::info;

type Error = Box<dyn std::error::Error + Send + Sync>;
type PrefixContext<'a> = poise::PrefixContext<'a, Data, Error>;

#[command(prefix_command)]
pub async fn remix(
  ctx: PrefixContext<'_>,
  #[description = "How to remix the image"]
  #[rest]
  prompt: Option<String>,
) -> Result<(), Error> {
  // As AI models take some time to respond, defer must be called first in order to keep
  // the interaction alive.
  ctx.defer().await?;
  
  let prompt = match prompt {
    Some(prompt) => prompt,
    None => {
      info!("No prompt provided for remix command");
      let embed = CreateEmbed::default()
        .title("Error")
        .description("Please instructions on how to remix.")
        .color(Color::from_rgb(137, 220, 235));
      ctx
        .send(CreateReply::default().embed(embed).reply(true))
        .await?;
      return Ok(());
    }
  };
  info!("Remix command invoked with prompt: {}", prompt);

  info!("Getting reply message...");
  let reply_message = match &ctx.msg.referenced_message {
    Some(message) => message,
    None => {
      let embed = CreateEmbed::default()
        .title("Error")
        .description("Please reply to a message with an image to remix it.")
        .color(Color::from_rgb(137, 220, 235));
      ctx
        .send(CreateReply::default().embed(embed).reply(true))
        .await?;
      return Ok(());
    }
  };
  info!("Got reply message");

  info!("Getting first image from message attachments...");
  let image =
    reply_message
      .attachments
      .iter()
      .find(|attachment| match &attachment.content_type {
        Some(content_type) => content_type.starts_with("image/"),
        None => false,
      });

  let image = match image {
    Some(image) => image,
    None => {
      let embed = CreateEmbed::default()
        .title("Error")
        .description("The referenced message does not contain an image.")
        .color(Color::from_rgb(137, 220, 235));
      ctx
        .send(CreateReply::default().embed(embed).reply(true))
        .await?;
      return Ok(());
    }
  };
  info!("Got first image from message attachments");

  let mime_type = image.content_type.clone().unwrap();
  let image_data = match image.download().await {
    Ok(image_data) => image_data,
    Err(_) => {
      info!("Failed to download");
      let embed = CreateEmbed::default()
        .title("Error")
        .description("Failed to retrieve the image from the referenced message.")
        .color(Color::from_rgb(137, 220, 235));
      ctx
        .send(CreateReply::default().embed(embed).reply(true))
        .await?;
      return Ok(());
    }
  };

  let gemini_imagen_client = &ctx.data().gemini_imagen_client;
  let imagen_prompt_with_image = GeminiImagenPromptWithImage {
    prompt: prompt.clone(),
    image_mime_type: mime_type,
    image_data,
  };

  let image_data = match gemini_imagen_client
    .prompt_with_image(imagen_prompt_with_image)
    .await
  {
    Ok(generated_image) => generated_image.image_data,
    Err(e) => {
      info!("Error remixing image: {:?}", e);
      let embed = CreateEmbed::default()
        .title("Error remixing image")
        .description(
          "Failed to generate the remixed image. Try altering the remix prompt.",
        )
        .color(Color::from_rgb(137, 220, 235));
      ctx
        .send(CreateReply::default().embed(embed).reply(true))
        .await?;
      return Ok(());
    }
  };

  let attachment = CreateAttachment::bytes(image_data, "image.png");

  let embed = CreateEmbed::default()
    .title("Remix!")
    .description(prompt)
    .color(Color::from_rgb(137, 220, 235))
    .attachment(&attachment.filename);

  ctx
    .send(
      CreateReply::default()
        .embed(embed)
        .attachment(attachment)
        .reply(true),
    )
    .await?;

  Ok(())
}

File: src/commands/remix.ts
import { AttachmentBuilder, EmbedBuilder, Message } from 'discord.js'
import { PrefixCommand } from '../models/commands'
import { ColorProvider } from '../utils/color-provider'
import {
  downloadDiscordImage,
  DownloadedImage,
  ImageDownloadError,
  MAX_REMIX_IMAGE_BYTES
} from '../utils/image-downloader'
import { ImageClient } from '../utils/image-client'
import { logger } from '../utils/logger'

export const createRemixCommand = (
  imageClient: ImageClient,
  colorProvider: ColorProvider
): PrefixCommand => ({
  name: 'remix',
  async execute (message) {
    const prompt = parseRemixPrompt(message.content)
    logger.info({ promptLength: prompt.length }, 'Processing remix command')

    if (!prompt) {
      await replyWithError(message, colorProvider, 'Please provide instructions on how to remix.')
      return
    }

    if (!message.reference?.messageId) {
      await replyWithError(message, colorProvider, 'Please reply to a message with an image to remix it.')
      return
    }

    try {
      const replyMessage = await message.channel.messages.fetch(message.reference.messageId)
      const image = await getFirstUsableImage(replyMessage)

      if (!image) {
        await replyWithError(message, colorProvider, 'Could not find a Discord-hosted image in the referenced message.')
        return
      }

      const response = await imageClient.editImage({
        prompt,
        imageMimeType: image.mimeType,
        imageData: image.buffer
      })

      const attachment = new AttachmentBuilder(response, { name: 'image.png' })
      await message.reply({ files: [attachment] })
    } catch (error) {
      logger.error({ error }, 'Error in remix command')
      await replyWithError(message, colorProvider, getRemixErrorDescription(error))
    }
  }
})

export const getFirstUsableImage = async (message: Message): Promise<DownloadedImage | undefined> => {
  const candidates: Array<{ source: 'attachment' | 'embed'; url: string }> = []
  for (const attachment of message.attachments.values()) {
    candidates.push({ source: 'attachment', url: attachment.url })
  }

  for (const embed of message.embeds) {
    if (embed.image?.proxyURL) {
      candidates.push({ source: 'embed', url: embed.image.proxyURL })
    }
  }

  let lastError: unknown
  for (const candidate of candidates) {
    try {
      return await downloadDiscordImage(candidate.url)
    } catch (error) {
      lastError = error
      logger.info({
        source: candidate.source,
        errorCode: error instanceof ImageDownloadError ? error.code : 'unknown'
      }, 'Unable to use remix image candidate')
    }
  }

  if (lastError) throw lastError
  return undefined
}

export const parseRemixPrompt = (content: string): string =>
  content.replace(/^!remix(?:\s+|$)/i, '').trim()

const replyWithError = async (
  message: Message,
  colorProvider: ColorProvider,
  description: string
): Promise<void> => {
  const embed = new EmbedBuilder()
    .setTitle('Error')
    .setDescription(description)
    .setColor(colorProvider.getErrorColor())
  await message.reply({ embeds: [embed] })
}

const getRemixErrorDescription = (error: unknown): string => {
  if (!(error instanceof ImageDownloadError)) {
    return 'There was an error processing the remix command.'
  }

  switch (error.code) {
    case 'unsupported_format':
      return 'Only PNG, JPEG, and WebP images can be remixed.'
    case 'too_large':
      return `The source image is too large. Use an image under ${MAX_REMIX_IMAGE_BYTES / 1024 / 1024} MiB.`
    case 'disallowed_url':
      return 'Only Discord-hosted or Discord-proxied images can be remixed.'
    case 'too_many_redirects':
    case 'download_failed':
      return 'The source image could not be downloaded from Discord.'
  }
}
